tutorials/haskell/golden-testing-in-haskell-with-hspec-gold/tutorial.md

```diff
@@ -0,0 +1,159 @@
+
+##The Main idea
+A golden test is just an **IO action** that writes its result to a file.
+To pass the test, the output **file** should be identical to the corresponding golden file,
+that contains the correct result for the given test.
+
+####‘A Golden File is the absolute source of truth. You validate your response against the Golden File. In a nutshell, the Golden file contains the response you expect from your test’
+
+##Changes on test flow
+The workflow for running the tests now becomes the following:
+
+* Run a test and make sure it fails.
+* Review the response from your test to see if it is accurate.
+* Run go test -update to update the existing golden file (or create new ones).
+* Re-run go test to ensure the test does not fail now.
+
+##How do Golden tests differentiate from unit tests?
+* If the output is larger in size, it's not practical to put it in the source code.
+
+* There is no need to escape quotes or binary data in the output.
+
+* When you add a new test, your testing framework can generate the missing golden file from the current output of the function.
+
+* It is best if you can write down the expected output without looking at the actual output, but it is not always possible.
+
+* The expected output can be automatically updated.
+
+* You can tell your test framework to update all golden files from the current outputs, then check git diff to ensure that all changes are valid, and commit.
+
+* If some of your tests suddenly started failing, you can use diff or other such tools to compare the golden file to the actual file and figure out what exactly changed.
+
+##What is Hspec-Golden?
+Hspec-Golden is a superset for the popular Haskell B.D.D. testing framework Hspec,
+that will allow us to create golden test for our Haskell projects,
+currently being developed by [Cristhian Motoche](https://github.com/CristhianMotoche) an active Open Source code enthusiast and Stackbuilders developer.
+
+##What do we need?
+For this tutorial, we need an environment running Haskell with base 4.6 up to 5,
+if we don’t have a Haskell environment we could refer to this [link](https://wiki.haskell.org/Haskell_in_5_steps#Install_Haskell) to get a clean installation using `stack` or `cabal`.
+
+We also need to have added as a dependency to our Haskell project the Hspec testing framework.
+
+With stack:
+
+``$ stack install hspec``
+
+With cabal:
+
+``$ cabal update && cabal install hspec``
+
+##Getting started
+This code sample will help us to understand how does basic golden testing works.
+
+##Installing the library:
+To install the hspec-golden superset we need to run this command on the console:
+
+``$ stack install hspec-golden``
+
+##Importing the libraries
+In our project on the Spec.hs file we will add the following lines.
+``` 
+import Test.Hspec
+import Test.Hspec.Golden
+
+main :: IO ()
+main = hspec $ do
```

Comment:
@MedAdrian Could you please include the code example as we've done with other [posts](https://github.com/stackbuilders/tutorials/tree/tutorials/tutorials/haskell/csv-encoding-decoding)?
```
tutorials/haskell/
    - code/
         src/
         test/
         ...
    - tutorial.md
```

Reply:
I've added a [repo](https://github.com/MedAdrian/hspec-golden-tutorial) for the code, and on the new version i append a link to it. (it's needed even with the code on a repo?)